### PR TITLE
backoffice

### PR DESCRIPTION
**Pull request by gqkc**
- Creation de la partie back office 
- Liste des users
- script to populate user

### CODE REVIEW

File: backoffice/www/handlers/base.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+class BaseBOHandler(tornado.web.RequestHandler):
+  """A base class for handlers."""
+
+  COOKIE = 'id'
+
+  def get_template_path(self):
+    return 'backoffice/www/templates/'
+
+
```
Comment (fpquintao) [inline backoffice/www/handlers/base.py]: it seems like we have more blank lines here than needed (?)
3 here, 1 between the two methods below.

File: backoffice/www/handlers/base.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import tornado.locale
+import tornado.web
+
+
+class BaseBOHandler(tornado.web.RequestHandler):
+  """A base class for handlers."""
+
+  COOKIE = 'id'
```
Comment (fpquintao) [inline backoffice/www/handlers/base.py]: not used?

File: backoffice/www/handlers/home.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+from backoffice.www.handlers.base import BaseBOHandler
+
+
+class HomeBOHandler(SessionMixin, BaseBOHandler):
+  ROUTE = '/'
+
+  @tornado.web.authenticated
+  def get(self):
+    self.render("home.html")
```
Comment (fpquintao) [inline backoffice/www/handlers/home.py]: github complanining of "no newline at the end of the file".

File: backoffice/model/encoder.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+from sqlalchemy.inspection import inspect
+
+class Serializer(object):
+
+    def serialize(self):
+        return {c: getattr(self, c) for c in inspect(self).attrs.keys()}
+
+    @staticmethod
+    def serialize_list(l):
+        return [obj.serialize() for obj in l]
```
Comment (fpquintao) [inline backoffice/model/encoder.py]: github complaining of missing line at the end of of the file.

File: backoffice/model/user.py [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+Base = declarative_base()
+
+
+class User(Base, Serializer):
+  __tablename__ = 'user'
+  id = Column(Integer, primary_key=True)
+  name = Column('name', String(255))
+
+  def to_json(self):
+    return {c.name: getattr(self, c.name) for c in self.__table__.columns}
```
Comment (fpquintao) [inline backoffice/model/user.py]: missing new line.

File: backoffice/www/handlers/login.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+  ROUTE= "/login"
+
+  def get(self):
+    self.write('<html><body><form action="/login" method="post">'
+               'Name: <input type="text" name="name">'
+               '<input type="submit" value="Sign in">'
+               '</form></body></html>')
+
+  def post(self):
+    self.set_secure_cookie("user", self.get_argument("name"))
```
Comment (fpquintao) [inline backoffice/www/handlers/login.py]: so users can always login? We don't even check if the user exists?

Fine for me, it's more out of curiosity; if this is the case, we don't even need a login handler. I guess the server will be running under a service that is only reachable by a few people who have the IP whitelisted in Amazon. I think this is a reasonable way to start.

File: backoffice/www/handlers/user.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    self.write('{} users so far!'.format(count))
+
+
+class UserJson(SessionMixin, BaseBOHandler):
+  ROUTE = '/users_json'
+
+  @tornado.web.authenticated
+  def get(self):
+    users = self.session.query(User).all()
+    # self.write({"users": users},)
```
Comment (fpquintao) [inline backoffice/www/handlers/user.py]: remove commented line?

File: backoffice/www/server.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    self.db = SQLAlchemy("sqlite+pysqlite:///{}".format(self.config.db.sqlite_path))
+    self.make_app()
+
+  def add_handler(self, handler, **kwargs):
+    route = os.path.join("/", handler.ROUTE)
+    self.routes.append((route, handler, kwargs))
+    logging.info("{} serving on {}".format(handler.__name__, route))
+
+  def make_app(self):
+    self.add_handler(LoginHandler)
```
Comment (fpquintao) [inline backoffice/www/server.py]: super nice.

File: backoffice/www/server.py [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    self.add_handler(ListUserHandler)
+    self.add_handler(UserJson)
+    self.add_handler(CreateUserHandler)
+    self.add_handler(HomeBOHandler)
+
+  def run(self):
+    logging.info(
+      "Running {} on port {}".format(self.__class__.__name__, self.port)
+    )
+    engine = create_engine("sqlite+pysqlite:///{}".format(self.config.db.sqlite_path), encoding='utf-8', echo=True)
```
Comment (fpquintao) [inline backoffice/www/server.py]: can we have in the constructor:

self.config ....
...
self.sqlite_path = "sqlite+pysqlite:///{}".format(self.config.db.sqlite_path)

then we can use it to initialise self.db (line 23) and the engine here.

File: backoffice/www/server.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+      "login_url": "/login",
+      "static_path": "backoffice/www/static",
+      "db": self.db
+    }
+    tornado.locale.load_translations('icubam/www/translations')
+    app = tornado.web.Application(self.routes, **settings)
+    app.listen(self.port)
+
+    io_loop = tornado.ioloop.IOLoop.current()
+    # for callback_obj in self.callbacks:
```
Comment (fpquintao) [inline backoffice/www/server.py]: should we drop these commented lines?

File: backoffice/www/static/list_users.js [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+$(document).ready(function() {
+  $('#example').DataTable({
+    ajax: {
+      url: '/users_json',
+      data: function(d){
+            console.log(d)
+            return d
+        },
+    },
+    /*columnDefs: [ {
```
Comment (fpquintao) [inline backoffice/www/static/list_users.js]: drop commented lines?

File: backoffice/www/handlers/user.py [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+    self.write({"data": User.serialize_list(users)})
+
+
+class ListUserHandler(SessionMixin, BaseBOHandler):
+  ROUTE = '/list_users'
+
+  @tornado.web.authenticated
+  def get(self):
+    self.render("list_users.html")
```
Comment (fpquintao) [inline backoffice/www/handlers/user.py]: I feel like this is missing the data? It's just invoking the template but we should pass the list of users that currently exist, like it's done in the /users_json handler (which I believe it's just for debug, which is great to have)

File: backoffice/www/handlers/home.py [hunk 13]
```diff
@@ -0,0 +1,12 @@
+import tornado
```
Comment (fpquintao) [inline backoffice/www/handlers/home.py]: I guess the backoffice should live inside the icubam folder.

File: backoffice/www/handlers/login.py (hunk 6 above)
Comment (rprimet) [reply backoffice/www/handlers/login.py]: Hmmm I'm out of the loop but that might warrant confirmation by @tehf0x and/or @olivierteboul?

File: scripts/populate_user.py [hunk 14]
```diff
@@ -0,0 +1,31 @@
+from absl import app
+from absl import flags
+from sqlalchemy import create_engine
```
Comment (fpquintao) [inline scripts/populate_user.py]: missing sqlachemy in the requirements.txt file?

**Comment by gqkc** [comment]
Sorry, I resolved a conversation unintentionally 
I said we want to check if the user exists of course. This will be open to external users accepted by the admin.

**Comment by gqkc** [comment]
I added a script to populate users in scripts

**Comment by dulacarnold** [comment]
This has been cut up into other PRs.

**Pull request closed without merging**